### ISSUE TITLE
AE2 export bus drivers and item NBT handling

### ISSUE THREAD

**Issue by fnuecke** [MightyPirates/OpenComputers#597]
Add a driver for AE2 export buses that allows
- configuring the export bus (i.e. set the 'filter'), using an item stack representation gotten from an inventory controller or the ME controller driver.
- trigger an export operation, optionally allowing to specify into which slot of the target inventory to try and export into.

Now to go off on a tangent a bit, this kind of touches on [this issue](https://github.com/MightyPirates/OpenComponents/issues/30), as the stack converter should then also retain the NBT info on stacks in some form (because there are many mod items that differentiate their type purely based on NBT data).

I'd like to avoid exposing the "plain" NBT data to Lua, because I feel this is too much 'implementation detail'. What I'm currently considering is wrapping the NBT in userdata. The downside of that is that it couldn't be... saved to disk, basically. Nor transmitted over a network. An alternative might be to return a plain UUID and keep track of NBT compounds in some magical hashmap that maps these UUIDs to the actual tag. That has the downside that the map'll never shrink, because it'd be impossible to tell when an entry isn't needed any longer. Unless we give the tags some kind of... time to live. Which would be very illogical from an ingame point of view.

Ideas and feedback on this would be very welcome!

**Comment by natedogith1** [comment]
Could you use encrypted (and maybe compressed) versions of the nbt data for lua-side representation?

**Comment by fnuecke** [comment]
Hmm, yeah, I guess that would work! Probably both, i.e. encrypt because otherwise it could be plainly available anyway (probably using a per-world automatically generated key) and compressed to avoid eating up too much RAM on Lua side, where memory is a relatively sparse resource, after all.

Hmm, does anyone know whether gzip compression is deterministic? I.e. would pushing the same data multiple times, leading to multiple compression runs, lead to the same output, so that it could be easily checked for equality?

**Comment by Xfel** [comment]
This is unrelated to NBT, but to the export bus thing: What would happen if I have multiple export buses in one tile? which one would be picked? I'd suggest adapting the interface block instead, that would be much clearer.

**Comment by fnuecke** [comment]
Oh right, multipart stuff. Guess I would have stumbled over that once I looked into implementing it sooner or later anyway. Hmm. Actually, I think this isn't so bad. Just provide the direction into which to export and an export bus will be chosen accordingly, if present. If none exists in that direction the operation will fail.

Edit: as for the NBT stuff... GZIP seems deterministic, encrypting on the other hand... unless using a _really_ simple scheme, in which case it's pointless, also. So I guess I'll just leave out the encryption. If you have GZIP in pure Lua, good for you. The other advantage of not encrypting it is that it's easier for others to decode it again on the Java side without OC having to make a decrypt method or the key itself available via the API.

**Comment by Kubuxu** [comment]
@magik6k it is gzip's specification reading time.
@fnuecke IMHO it is great as you will be able to check for example TiC tool durability or mob in safari net.

**Comment by fnuecke** [comment]
Done. NBT stuff is now added to tables representing item stacks as a byte array of the gzipped data (which is just a string in Lua), can be disabled in config if you're paranoid.

To make converting these tables back to item stacks easier, I added convenience methods for that to the Arguments interface.

The export bus driver is also done. It allows getting and setting the item configuration of export buses next to an adapter (the actual bus is chosen by passing along the side to the callbacks), as well as triggering a single export operation into a specific slot. This will still respect sided inventories and all that - in general it should behave almost exactly like a normal export operation (i.e. it'll also drain energy from the ME network, as well as respect fuzzy and speed upgrades), except that the target slot can be chosen. To make proper us of this, you'll probably want to disable default exporting, e.g. by installing an AE redstone card and setting it to requiring a high signal or so.

Edit: oh, and regarding safari nets, there's a driver for that since a few days a go, and TiC tool durability works using the normal damage values for me (i.e. stack.damage).
